src/main/java/me/crespel/runtastic/converter/ExportConverter.java

```diff
@@ -9,60 +9,125 @@
 
-	public List<SportSession> listSportSessions(File path) throws FileNotFoundException, IOException {
-		path = normalizeSportSessionPath(path);
+	public List<SportSession> listSportSessions(File path, boolean full) throws FileNotFoundException, IOException {
+		path = normalizeExportPath(path, SPORT_SESSIONS_DIR);
 		List<SportSession> sessions = new ArrayList<>();
 		File[] files = path.listFiles(file -> file.getName().endsWith(".json"));
 		for (File file : files) {
-			sessions.add(parser.parseSportSession(file));
+			sessions.add(parser.parseSportSession(file,full));
 		}
 
 		Collections.sort(sessions);
 		return sessions;
 	}
 
+	public User getUser(File path) throws FileNotFoundException, IOException {
+		return parser.parseUser(new File(normalizeExportPath(path, USER_DIR), "user.json"));
+	}
+
 	public SportSession getSportSession(File path, String id) throws FileNotFoundException, IOException {
-		path = normalizeSportSessionPath(path);
-		return parser.parseSportSession(new File(path, id + ".json"));
+		return parser.parseSportSession(new File(normalizeExportPath(path, SPORT_SESSIONS_DIR), id + ".json"), true);
+	}
+
+	public SportSession getSportSessionWithPhoto(File path, String photoid) throws FileNotFoundException, IOException {
+		String sessionid = null;
+
+		File photofile = new File(normalizeExportPath(path, PHOTOS_DIR), photoid + ".jpg");
+		if( photofile.exists() ) {
+			// photo file found ...
+
+			ImagesMetaData image = parser.parseImagesMetaData(new File(normalizeExportPath(path, PHOTOS_META_DATA_DIR), photoid + ".json"));
+			if( image != null ) {
+				// photo meta data file found ...
+
+				// search trough sport session album data, to find sport session related to the photo
+				File[] files = normalizeExportPath(path, PHOTOS_SPORT_SESSION_ALBUMS_DIR).listFiles(file -> file.getName().endsWith(".json"));
+				for( File file : files ) {
+					try {
+						SportSessionAlbums mysessionalbum = parser.parseSportSessionAlbumsData(file);
+						if (mysessionalbum.getPhotosIds().contains(photoid)) {
+							// Sport session id found within sport session albums
+							sessionid = mysessionalbum.getId();
+						}
+					} catch (IOException e) {
+						throw new RuntimeException(e);
+					}
+				}
+			}
+		}
+
+		return getSportSession(path, sessionid);
+	}
+
+	public List<SportSession> convertSportSessions(File path, String format) throws FileNotFoundException, IOException {
+		File[] files = normalizeExportPath(path, SPORT_SESSIONS_DIR).listFiles(file -> file.getName().endsWith(".json"));
+		List<SportSession> sessionlist = new ArrayList<>();
+		Arrays.asList(files).parallelStream().forEach(file -> {
+			try {
+				SportSession session = parser.parseSportSession(file, true);
+				if (session.getGpsData() != null || session.getHeartRateData() != null || session.getGpx() != null) {
+					GpxType gpx = (GpxType) mapper.mapSportSession(session, format);

```

Comment:
When converting with format="tcx", the mapper will NOT return a `GpxType` and this will fail. Additionally, if `session.getGpx()` is not null, there should be no need to convert to GPX since it's already done...?

Reply:
Fixed, for "gpx". The udpated "gpx" data is now stored back to session in "mapSportSession" function. So there is no need to evaluate return parameter anymore in "convertSportSessions" method.

It is still needed to "map" data even if "gpx" data is already there. If it exists already, it cotains the loaded data from (new) export format. The "map" method extracts the tracking points and adds them as new "gpx" data (with own meta data etc.). Because of that the "map" needs to be executed.

I just noticed, that I'm not sure that "tcx" is correct handled in case the export contains new "gpx" data! Do we need a "gpx" -> "tcx" converter?